File: uos_iot_client_nodejs/uos_iot_client.js
#!/usr/bin/env node
/**
 * UrbanOS IoT MQTT Client (Node.js)
 * - Publishes ONLY: lat, lon, elevation, speed, activity, timestamp
 * - Topic: client/{client_id}/session/{session_id}/
 * - QoS 1, no retain, offline queue, optional 26h session rotation
 */

const mqtt = require("mqtt");
const { randomUUID } = require("crypto");

// ──────────────────────────────────────────────────────────────────────────────
// Config via env
// ──────────────────────────────────────────────────────────────────────────────
const MQTT_BROKER = process.env.MQTT_BROKER || "mqtt://localhost:1883";
const MQTT_USERNAME = process.env.MQTT_USERNAME || undefined;
const MQTT_PASSWORD = process.env.MQTT_PASSWORD || undefined;
const MQTT_QOS = parseInt(process.env.MQTT_QOS || "1", 10);
const MQTT_KEEPALIVE = parseInt(process.env.MQTT_KEEPALIVE || "60", 10);

const CLIENT_ID = process.env.UOS_CLIENT_ID || `cli-${randomUUID().slice(0, 12)}`;
const SESSION_TTL_HOURS = parseInt(process.env.UOS_SESSION_TTL_HOURS || "26", 10);
const TOPIC_TEMPLATE = (process.env.UOS_TOPIC_TEMPLATE || "client/{client_id}/session/{session_id}/").replace(/\/+$/, "") + "/";

const PUBLISH_INTERVAL = parseFloat(process.env.UOS_PUBLISH_INTERVAL || "1.0"); // seconds
const MAX_QUEUE = parseInt(process.env.UOS_MAX_QUEUE || "10000", 10);

// ──────────────────────────────────────────────────────────────────────────────
// Session (stable id per ~26h window)
// ──────────────────────────────────────────────────────────────────────────────
function nowUtc() { return new Date(); }

class Session {
  constructor(clientId) {
    this.clientId = clientId;
    this.sessionId = process.env.UOS_SESSION_ID || randomUUID();
    this.startedAt = nowUtc();
  }
  maybeRotate() {
    if (!SESSION_TTL_HOURS) return;
    const ageMs = nowUtc() - this.startedAt;
    if (ageMs >= SESSION_TTL_HOURS * 3600 * 1000) {
      console.log("🔄 Rotating session_id after TTL");
      this.sessionId = randomUUID();
      this.startedAt = nowUtc();
    }
  }
}

// ──────────────────────────────────────────────────────────────────────────────
// Offline queue
// ──────────────────────────────────────────────────────────────────────────────
class OfflineQueue {
  constructor(limit) {
    this.limit = limit;
    this.buf = [];
  }
  enqueue(msg) {
    if (this.buf.length >= this.limit) {
      // drop oldest
      this.buf.shift();
      console.warn("🧹 Queue full → dropped oldest");
    }
    this.buf.push(msg);
  }
  drain(publishFn) {
    let sent = 0;
    while (this.buf.length) {
      const msg = this.buf.shift();
      publishFn(msg);
      sent++;
    }
    if (sent) console.log(`✅ Drained ${sent} queued messages`);
  }
}

// ──────────────────────────────────────────────────────────────────────────────
// Publisher
// ──────────────────────────────────────────────────────────────────────────────
class UOSPublisher {
  constructor() {
    this.session = new Session(CLIENT_ID);
    this.queue = new OfflineQueue(MAX_QUEUE);

    const options = {
      keepalive: MQTT_KEEPALIVE,
      username: MQTT_USERNAME,
      password: MQTT_PASSWORD,
      reconnectPeriod: 1000,      // ms
      resubscribe: false,
      clean: true,
      clientId: `uos-${CLIENT_ID}-${randomUUID().slice(0, 6)}`,
    };

    this.client = mqtt.connect(MQTT_BROKER, options);
    this.client.on("connect", () => {
      console.log("✅ Connected to MQTT broker");
      this.queue.drain((m) => this._publishRaw(m));
    });
    this.client.on("reconnect", () => console.warn("↻ Reconnecting…"));
    this.client.on("offline", () => console.warn("⚠ MQTT offline"));
    this.client.on("error", (err) => console.error("❌ MQTT error:", err.message));
    this.client.on("close", () => console.warn("⚠ MQTT connection closed"));
  }

  _topic() {
    return TOPIC_TEMPLATE
      .replace("{client_id}", this.session.clientId)
      .replace("{session_id}", this.session.sessionId);
  }

  publishPoint({ lat, lon, elevation = null, speed = null, activity = null, timestamp = null }) {
    this.session.maybeRotate();

    const ts = (timestamp instanceof Date ? timestamp : nowUtc()).toISOString();
    const payload = {
      lat: Number(lat),
      lon: Number(lon),
      elevation: elevation == null ? null : Number(elevation),
      speed: speed == null ? null : Number(speed),
      activity: activity || null,
      timestamp: ts,
    };

    const msg = { topic: this._topic(), payload };
    if (this.client.connected) {
      this._publishRaw(msg);
    } else {
      this.queue.enqueue(msg);
    }
  }

  _publishRaw({ topic, payload }) {
    const data = JSON.stringify(payload);
    this.client.publish(topic, data, { qos: MQTT_QOS, retain: false }, (err) => {
      if (err) {
        console.warn("📥 Offline queueing due to publish error:", err.message);
        this.queue.enqueue({ topic, payload });
      }
    });
  }

  close() {
    try { this.client.end(true); } catch {}
  }
}

// ──────────────────────────────────────────────────────────────────────────────
// Example sensor (replace with your GNSS/IMU feed)
// ──────────────────────────────────────────────────────────────────────────────
async function* fakeSensorStream() {
  let lat = 59.3293, lon = 18.0686, elev = 10.0, speed = 1.2, activity = "walking";
  while (true) {
    yield { lat, lon, elevation: elev, speed, activity };
    lat += 0.00005; lon += 0.00007;
    await new Promise(r => setTimeout(r, PUBLISH_INTERVAL * 1000));
  }
}

// ──────────────────────────────────────────────────────────────────────────────
async function main() {
  const pub = new UOSPublisher();

  const stop = (() => {
    let s = false;
    const handler = () => { s = true; };
    process.on("SIGINT", handler);
    process.on("SIGTERM", handler);
    return () => s;
  })();

  try {
    for await (const reading of fakeSensorStream()) {
      if (stop()) break;
      pub.publishPoint(reading); // replace with real device data
    }
  } finally {
    pub.close();
  }
}

if (require.main === module) main();

File: uos_iot_client_python/uos_iot_client.py
#!/usr/bin/env python3
"""
UrbanOS IoT MQTT Client
- Publishes geodata messages to a private topic: client/{client_id}/session/{session_id}/
- Sends ONLY: lat, lon, elevation, speed, activity, timestamp (UTC ISO8601)
- Queues locally on disconnect and drains when reconnected (self-healing)
- Safe defaults: QoS 1 (at-least-once), no retain, no PII
"""

import os
import json
import time
import uuid
import queue
import signal
import logging
from datetime import datetime, timezone
from typing import Optional, Dict, Any

import paho.mqtt.client as mqtt

# ──────────────────────────────────────────────────────────────────────────────
# Configuration (env overrides)
# ──────────────────────────────────────────────────────────────────────────────
MQTT_BROKER = os.getenv("MQTT_BROKER", "localhost")
MQTT_PORT = int(os.getenv("MQTT_PORT", "1883"))
MQTT_USERNAME = os.getenv("MQTT_USERNAME")      # optional
MQTT_PASSWORD = os.getenv("MQTT_PASSWORD")      # optional
MQTT_KEEPALIVE = int(os.getenv("MQTT_KEEPALIVE", "60"))
MQTT_QOS = int(os.getenv("MQTT_QOS", "1"))      # QoS 1 pairs with broker persistence

CLIENT_ID = os.getenv("UOS_CLIENT_ID", f"cli-{uuid.uuid4().hex[:12]}")
SESSION_TTL_HOURS = int(os.getenv("UOS_SESSION_TTL_HOURS", "26"))

# Topic template; deployer can switch formats but must keep this shape
TOPIC_TEMPLATE = os.getenv(
    "UOS_TOPIC_TEMPLATE",
    "client/{client_id}/session/{session_id}/"
).rstrip("/") + "/"

# Send interval seconds (device sampling)
PUBLISH_INTERVAL = float(os.getenv("UOS_PUBLISH_INTERVAL", "1.0"))

# Local queue caps (protect memory in prolonged outages)
MAX_QUEUE = int(os.getenv("UOS_MAX_QUEUE", "10000"))

logging.basicConfig(
    level=getattr(logging, os.getenv("LOG_LEVEL", "INFO")),
    format="%(asctime)s | %(levelname)s | %(message)s"
)

# ──────────────────────────────────────────────────────────────────────────────
# Session management
# ──────────────────────────────────────────────────────────────────────────────
def now_utc() -> datetime:
    return datetime.now(timezone.utc)

class Session:
    """
    One 26h logical session per UrbanOS spec.
    Devices DO NOT need to compute start/end; UrbanOS will window by timestamp.
    We still maintain a stable session_id to avoid per-message UUID churn.
    """
    def __init__(self, client_id: str):
        self.client_id = client_id
        # Stable session_id per 26h window; deployer can rotate on their schedule
        self.session_id = os.getenv("UOS_SESSION_ID") or self._new_session_id()
        self.start_time = now_utc()

    def _new_session_id(self) -> str:
        # Using a compact, opaque string (not PII). Could also be numeric.
        return str(uuid.uuid4())

    def maybe_rotate(self):
        # Optional client-side rotation. UrbanOS works even if you never rotate.
        if (now_utc() - self.start_time).total_seconds() >= SESSION_TTL_HOURS * 3600:
            logging.info("🔄 Rotating session_id after TTL")
            self.session_id = self._new_session_id()
            self.start_time = now_utc()

# ──────────────────────────────────────────────────────────────────────────────
# Local queue for offline persistence
# ──────────────────────────────────────────────────────────────────────────────
class OfflineQueue:
    def __init__(self, maxsize: int = MAX_QUEUE):
        self.q = queue.Queue(maxsize=maxsize)

    def put(self, item: Dict[str, Any]):
        try:
            self.q.put_nowait(item)
            return True
        except queue.Full:
            # Drop oldest; keep stream fresh
            try:
                _ = self.q.get_nowait()
                self.q.put_nowait(item)
                logging.warning("🧹 Queue full → dropped oldest to admit new item")
                return True
            except Exception:
                logging.error("❌ Queue full and could not drop oldest")
                return False

    def drain_to(self, fn_publish):
        sent = 0
        while not self.q.empty():
            try:
                fn_publish(self.q.get_nowait())
                sent += 1
            except Exception as e:
                logging.error(f"❌ Failed while draining queue: {e}")
                # Put it back and stop draining to avoid spin
                self.q.task_done()
                break
            self.q.task_done()
        if sent:
            logging.info(f"✅ Drained {sent} queued messages")

# ──────────────────────────────────────────────────────────────────────────────
# MQTT client
# ──────────────────────────────────────────────────────────────────────────────
class UOSPublisher:
    def __init__(self):
        self.session = Session(CLIENT_ID)
        self.offline = OfflineQueue()
        self.client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2, client_id=f"uos-{CLIENT_ID}")
        self._configure()

    def _configure(self):
        self.client.enable_logger()
        if MQTT_USERNAME:
            self.client.username_pw_set(MQTT_USERNAME, MQTT_PASSWORD)

        # Robust reconnect posture
        self.client.reconnect_delay_set(min_delay=1, max_delay=60)
        self.client.max_inflight_messages_set(100)

        self.client.on_connect = self.on_connect
        self.client.on_disconnect = self.on_disconnect

        self.client.connect(MQTT_BROKER, MQTT_PORT, MQTT_KEEPALIVE)
        self.client.loop_start()

    # Callbacks
    def on_connect(self, client, userdata, flags, rc, properties=None):
        if rc == 0:
            logging.info("✅ Connected to MQTT broker")
            # Drain queued messages once we’re online
            self.offline.drain_to(self._publish_raw)
        else:
            logging.error(f"❌ MQTT connect failed (rc={rc})")

    def on_disconnect(self, client, userdata, rc):
        # rc 0 is clean; others mean unexpected disconnect
        if rc != 0:
            logging.warning(f"⚠ Disconnected from MQTT (rc={rc})")

    # Public API: publish one point
    def publish_point(self, lat: float, lon: float,
                      elevation: Optional[float] = None,
                      speed: Optional[float] = None,
                      activity: Optional[str] = None,
                      timestamp: Optional[datetime] = None):
        """
        Publish one geodata point. timestamp defaults to now (UTC).
        """
        self.session.maybe_rotate()
        ts = timestamp or now_utc()

        payload = {
            "lat": float(lat),
            "lon": float(lon),
            "elevation": None if elevation is None else float(elevation),
            "speed": None if speed is None else float(speed),
            "activity": activity or None,
            "timestamp": ts.isoformat(),
        }

        topic = TOPIC_TEMPLATE.format(
            client_id=self.session.client_id,
            session_id=self.session.session_id
        )

        msg = {"topic": topic, "payload": payload}

        # Try to send immediately; if not connected, queue
        try:
            self._publish_raw(msg)
        except Exception as e:
            logging.warning(f"📥 Offline queueing ({e}); will resend on reconnect")
            self.offline.put(msg)

    # Low-level send; raises if publish failed
    def _publish_raw(self, msg: Dict[str, Any]):
        info = self.client.publish(
            msg["topic"],
            json.dumps(msg["payload"], separators=(",", ":")),
            qos=MQTT_QOS,
            retain=False
        )
        # Block briefly until sent to local socket; non-blocking would also work
        info.wait_for_publish()
        if info.rc != mqtt.MQTT_ERR_SUCCESS:
            raise RuntimeError(f"publish rc={info.rc}")
        logging.debug(f"📤 {msg['topic']} {msg['payload']}")

    def close(self):
        self.client.loop_stop()
        self.client.disconnect()

# ──────────────────────────────────────────────────────────────────────────────
# Example data source (replace with real GNSS/IMU/SDK)
# ──────────────────────────────────────────────────────────────────────────────
def fake_sensor_stream():
    """
    Replace this generator with your device’s real location source.
    Yields dicts with lat/lon/elevation/speed/activity.
    """
    lat, lon = 59.3293, 18.0686  # Stockholm 🙂
    speed = 1.2
    elev = 10.0
    activity = "walking"
    while True:
        yield {
            "lat": lat,
            "lon": lon,
            "elevation": elev,
            "speed": speed,
            "activity": activity
        }
        # Drift a tiny bit
        lat += 0.00005
        lon += 0.00007
        time.sleep(PUBLISH_INTERVAL)

# ──────────────────────────────────────────────────────────────────────────────
# Main
# ──────────────────────────────────────────────────────────────────────────────
def main():
    pub = UOSPublisher()

    # Graceful shutdown
    stop = False
    def _sig(_a, _b):
        nonlocal stop
        stop = True
    signal.signal(signal.SIGINT, _sig)
    signal.signal(signal.SIGTERM, _sig)

    try:
        for reading in fake_sensor_stream():
            if stop:
                break
            pub.publish_point(
                lat=reading["lat"],
                lon=reading["lon"],
                elevation=reading.get("elevation"),
                speed=reading.get("speed"),
                activity=reading.get("activity"),
                # tim
